core/src/main/java/org/axonframework/commandhandling/gateway/GatewayProxyFactory.java

```diff
@@ -181,7 +191,7 @@ public GatewayProxyFactory(CommandBus commandBus, RetryScheduler retryScheduler,
             InvocationHandler dispatcher = new DispatchOnInvocationHandler(commandBus, retryScheduler,
                                                                            dispatchInterceptors, extractors,
                                                                            commandCallbacks, true);
-            if (!Future.class.equals(gatewayMethod.getReturnType())) {
+            if (!Future.class.isAssignableFrom(gatewayMethod.getReturnType())) {
```

Comment:
This will not allow declared return type of Completion stage to be used. The check should be:
`if(!gateway.getReturnType().isAssignableFrom(CompletableFuture.class))`. This checks if a CompletableFuture is a valid return value for the method. That's the case when the declared type is either CompletionStage, Future or CompletableFuture.